src/spec/expectations.cr

```diff
@@ -1,4 +1,58 @@
 module Spec
+  # A list of diff command candidates.
+  DIFF_COMMANDS = %w(gdiff diff)
+
+  # A diff command path to use in diff computation.
+  class_property diff_command : String? do
+    DIFF_COMMANDS.compact_map { |name| Process.find_executable(name) }.first?
+  end
+
+  # Compute the difference between two values *expected_value* and *actual_value*
+  # by using `diff` command.
+  def self.diff_values(expected_value, actual_value)
+    expected = expected_value.pretty_inspect
+    actual = actual_value.pretty_inspect
+
+    result = diff expected, actual
+
+    # When the diff output is nothing even though the two values do not equal,
+    # it returns a fallback message so far.
+    if result && result.empty?
+      klass = expected_value.class
+      return <<-MSG
+        No visible difference in the `#{klass}#pretty_inspect` output.
+        You should look at the implementation of `#==` on #{klass} or its members.
+        MSG
+    end
+
+    result
+  end
+
+  # Compute the difference between two strings *expected* and *actual*
+  # by using `diff` command.
+  def self.diff(expected, actual)
+    # If the diff command is available and outputs contain a newline,
+    # then it computes diff of them.
+    diff_command = Spec.diff_command
+    return unless diff_command && (expected.includes?('\n') || actual.includes?('\n'))
+
+    expected_file = File.tempfile("expected") { |f| f.puts expected }
```

Comment:
If this will be merged, there should be some investigation towards limiting Crystal's tmp file footprint to a particular subdirectory. Maybe there is already a spec-global top-level directory being used somewhere? I know in Crystal's own specs there is but that's not it.